core/services/ocr2/plugins/ocr2keeper/evm/registry.go

```diff
@@ -0,0 +1,771 @@
+	"github.com/smartcontractkit/chainlink/core/chains/evm/logpoller"
+	"github.com/smartcontractkit/chainlink/core/gethwrappers/generated/keeper_registry_wrapper2_0"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/pg"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+var (
+	ErrLogReadFailure              = fmt.Errorf("failure reading logs")
+	ErrHeadNotAvailable            = fmt.Errorf("head not available")
+	ErrRegistryCallFailure         = fmt.Errorf("registry chain call failure")
+	ErrBlockKeyNotParsable         = fmt.Errorf("block identifier not parsable")
+	ErrUpkeepKeyNotParsable        = fmt.Errorf("upkeep key not parsable")
+	ErrInitializationFailure       = fmt.Errorf("failed to initialize registry")
+	ErrContextCancelled            = fmt.Errorf("context was cancelled")
+	ErrABINotParsable              = fmt.Errorf("error parsing abi")
+	ActiveUpkeepIDBatchSize  int64 = 1000
+	separator                      = "|"
+	reInitializationDelay          = 3 * time.Minute
+)
+
+type LatestBlockGetter interface {
+	LatestBlock() int64
+}
+
+func NewEVMRegistryServiceV2_0(addr common.Address, client evm.Chain, lggr logger.Logger) (*EvmRegistry, error) {
+	abi, err := abi.JSON(strings.NewReader(keeper_registry_wrapper2_0.KeeperRegistryABI))
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrABINotParsable, err)
+	}
+
+	registry, err := keeper_registry_wrapper2_0.NewKeeperRegistry(addr, client.Client())
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to create caller for address and backend", ErrInitializationFailure)
+	}
+
+	r := &EvmRegistry{
+		HeadProvider: HeadProvider{
+			ht: client.HeadTracker(),
+			hb: client.HeadBroadcaster(),
+		},
+		lggr:     lggr,
+		poller:   client.LogPoller(),
+		addr:     addr,
+		client:   client.Client(),
+		registry: registry,
+		abi:      abi,
+		packer:   &evmRegistryPackerV2_0{abi: abi},
+		headFunc: func(types.BlockKey) {},
+		chLog:    make(chan logpoller.Log, 1000),
+	}
+
+	if err := r.registerEvents(addr); err != nil {
+		return nil, fmt.Errorf("logPoller error while registering automation events: %w", err)
+	}
+
+	return r, nil
+}
+
+var upkeepStateEvents = []common.Hash{
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepMigrated{}.Topic(),   // removes upkeep id and detail from registry
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepRegistered{}.Topic(), // adds new upkeep id to registry
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepReceived{}.Topic(),   // adds multiple new upkeep ids to registry
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepCheckDataUpdated{}.Topic(),
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepGasLimitSet{}.Topic(),
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepCanceled{}.Topic(),
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepPaused{}.Topic(),
+	keeper_registry_wrapper2_0.KeeperRegistryUpkeepUnpaused{}.Topic(),
+	keeper_registry_wrapper2_0.KeeperRegistryFundsAdded{}.Topic(),
```

Comment:
did we go through the log list to verify we are not missing any log here?

Reply:
Funds added can be removed.